executor/src/witgen/jit/block_machine_processor.rs

```diff
@@ -0,0 +1,375 @@
+            }
+        }
+
+        // Solve for the block witness.
+        // Fails if any machine call cannot be completed.
+        self.solve_block(&mut witgen)?;
+
+        // For each unknown argument, get the value from the expression in the connection's RHS.
+        // If assign() fails, it means that we weren't able to to solve for the operation's output.
+        for (index, expr) in connection.right.expressions.iter().enumerate() {
+            if !known_args[index] {
+                let param_i =
+                    AffineSymbolicExpression::from_known_symbol(Variable::Param(index), None);
+                witgen
+                    .assign(expr, self.latch_row as i32, param_i)
+                    .map_err(|_| format!("Could not solve for params[{index}]"))?;
+            }
+        }
+
+        Ok(witgen.code())
+    }
+
+    /// Repeatedly processes all identities on all rows, until no progress is made.
+    /// Fails iff there are incomplete machine calls in the latch row.
+    fn solve_block(
+        &self,
+        witgen: &mut WitgenInference<T, FixedEvaluatorForFixedData<T>>,
+    ) -> Result<(), String> {
+        let mut complete = HashSet::new();
+        for i in 0.. {
+            let mut progress = false;
+
+            // TODO: This algorithm is assuming a rectangular block shape.
+            for row in (0..self.block_size) {
+                for id in self.machine_parts.identities.iter() {
+                    if !complete.contains(&(id.id(), row)) {
+                        let result = witgen.process_identity(id, row as i32);
+                        if result.complete {
+                            complete.insert((id.id(), row));
+                        }
+                        progress |= result.progress;
+                    }
+                }
+            }
+            if !progress {
+                log::debug!("Finishing after {} iterations", i);
+                break;
+            }
+        }
+
+        // If any machine call could not be completed, that's bad because machine calls typically have side effects.
+        // So, the underlying lookup / permutation / bus argument likely does not hold.
+        // TODO: This assumes a rectangular block shape.
+        let has_incomplete_machine_calls = (0..self.block_size)
+            .flat_map(|row| {
+                self.machine_parts
+                    .identities
+                    .iter()
+                    .map(move |id| (id, row))
+            })
+            .filter(|(identity, row)| !complete.contains(&(identity.id(), *row)))
+            .any(|(identity, _row)| !is_machine_call(identity));
+
+        match has_incomplete_machine_calls {
+            true => Err("Incomplete machine calls".to_string()),
+            false => Ok(()),
+        }
+    }
+}
+
+fn is_machine_call<T>(identity: &Identity<T>) -> bool {
+    matches!(
+        identity,
+        Identity::Lookup(_)
```

Comment:
Maybe better to use a `match` without default so that we will get an error if we extend the enum?